hydrolib/core/io/ini/serializer.py

```diff
@@ -0,0 +1,246 @@
+from dataclasses import dataclass
+from functools import reduce
+from hydrolib.core.io.ini.models import CommentBlock, Document, Property, Section
+from hydrolib.core.basemodel import BaseModel
+from itertools import chain, count, repeat
+from pathlib import Path
+from typing import Any, Iterable, List, Optional, Sequence, Tuple, Union
+
+
+class SerializerConfig(BaseModel):
+    """SerializerConfig defines the configuration options of the Serializer
+
+    Attributes:
+        section_indent (int):
+            The number of spaces with which whole sections should be indented.
+            Defaults to 0.
+        property_indent (int):
+            The number of spaces with which properties should be indented relative to
+            the section header (i.e. the full indent equals the section_indent plus
+            property_indent). Defaults to 4.
+        datablock_indent (int):
+            The number of spaces with which datablock rows are indented relative to
+            the section header (i.e. the full indent equals the section_indent plus
+            datablock_indent). Defaults to 8.
+        datablock_spacing (int):
+            The number of spaces between datablock columns. Note that there might be
+            additional offset to ensure . is lined out. Defaults to 4.
+        comment_delimiter (str):
+            The character used to delimit comments.
+    """
+
+    section_indent: int = 0
+    property_indent: int = 4
+    datablock_indent: int = 8
+    datablock_spacing: int = 4
+    comment_delimiter: str = "#"
+
+    @property
+    def total_property_indent(self):
+        return self.section_indent + self.property_indent
+
+    @property
+    def total_datablock_indent(self):
+        return self.section_indent + self.datablock_indent
+
+
+@dataclass
+class _Lengths:
+    # Internal data class to store the lengths of the elements in a section.
+    max_key_length: int
+    max_value_length: int
+    max_datablock_element_length: Optional[Sequence[int]] = None
+
+    @staticmethod
+    def _calculate_max_length_datablock(
+        datablock: Optional[Sequence[Sequence[str]]],
+    ) -> Optional[Sequence[int]]:
+        if datablock is None or len(datablock) < 1:
+            return None
+
+        def folder(acc: Tuple, row: Sequence[str]) -> Tuple:
+            return tuple(max(curr_max, len(elem)) for curr_max, elem in zip(acc, row))
```

Comment:
This is a little hard for me to read :p

Reply:
it is a fold function, though in the python sense, calling it a reductor might have been a better name. Basically it says

we have our acc, which is short for accumulator, which accumulates the results, i.e. it contains the data at one point in the reduction, and a next row consisting of our sequence of strings, and we return the accumulator at the end of processing this row. If we then put this in the reduce function, with the list of rows and an initial value:

```python
reduce(folder, datablock, (0,) * len(datablock[0]))
```

It will basically for each row execute the new accumulator with the folder, and return those values.

The folder basically calculates the max value for each element in the row and the max length till then for elements in that column. Because both the acc and row have the same elements, say 4 for the sake of the argument, if we zip them together, we get pairs of the previous max size, and the new element. We then calculate for each pair the new max length for that pair by taking the max between the previous max size and the length of the new element. We stuff all these values in a tuple, and we get back our new tuple consisting of the max elements until the new element.

Reply:
If you need this amount of text to explain it, maybe add that to the code. Python code should be readable, so don't be afraid to use more lines. Make sure to explain the why/thoughts behind this design in there something, maybe even in the module docstring.